src/flamenco/snapshot/fd_snapshot_create.c

```diff
@@ -0,0 +1,1219 @@
+     in the cache is the same as the data in the accounts db.
+     
+     The other problem is that the Firedancer stakes cache does not evict old
+     entries and doesn't update delegations within the cache. The cache will
+     just insert new pubkeys as stake accounts are created/delegated to. To
+     make the cache conformant for the snapshot, old accounts should be removed
+     from the snapshot and all of the delegations should be updated. */
+
+  /* First populate the vote accounts using the vote accounts/stakes cache. 
+     We can populate over all of the fields except we can't reserialize the
+     vote account data. Instead we will copy over the raw contents of all of
+     the vote accounts. */
+
+  ulong vote_accounts_len                      = fd_vote_accounts_pair_t_map_size( old_stakes->vote_accounts.vote_accounts_pool, old_stakes->vote_accounts.vote_accounts_root );
+  new_stakes->vote_accounts.vote_accounts_pool = fd_vote_accounts_pair_serializable_t_map_alloc( snapshot_ctx->valloc, fd_ulong_max(vote_accounts_len, 15000 ) );
+  new_stakes->vote_accounts.vote_accounts_root = NULL;
+
+  for( fd_vote_accounts_pair_t_mapnode_t * n = fd_vote_accounts_pair_t_map_minimum(
+       old_stakes->vote_accounts.vote_accounts_pool,
+       old_stakes->vote_accounts.vote_accounts_root );
+       n;
+       n = fd_vote_accounts_pair_t_map_successor( old_stakes->vote_accounts.vote_accounts_pool, n ) ) {
+    
+    fd_vote_accounts_pair_serializable_t_mapnode_t * new_node = fd_vote_accounts_pair_serializable_t_map_acquire( new_stakes->vote_accounts.vote_accounts_pool );
+    new_node->elem.key   = n->elem.key;
+    new_node->elem.stake = n->elem.stake;
+    /* Now to populate the value, lookup the account using the acc mgr */
+    FD_BORROWED_ACCOUNT_DECL( vote_acc );
+    int err = fd_acc_mgr_view( snapshot_ctx->acc_mgr, NULL, &n->elem.key, vote_acc );
+    if( FD_UNLIKELY( err ) ) {
+      FD_LOG_WARNING(( "Failed to view vote account from stakes cache %s", FD_BASE58_ENC_32_ALLOCA(&n->elem.key) ));
+      return -1;
+    }
+
+    new_node->elem.value.lamports   = vote_acc->const_meta->info.lamports;
+    new_node->elem.value.data_len   = vote_acc->const_meta->dlen;
+    new_node->elem.value.data       = fd_valloc_malloc( snapshot_ctx->valloc, 8UL, vote_acc->const_meta->dlen );
+    fd_memcpy( new_node->elem.value.data, vote_acc->const_data, vote_acc->const_meta->dlen );
+    fd_memcpy( &new_node->elem.value.owner, &vote_acc->const_meta->info.owner, sizeof(fd_pubkey_t) );
+    new_node->elem.value.executable = vote_acc->const_meta->info.executable;
+    new_node->elem.value.rent_epoch = vote_acc->const_meta->info.rent_epoch;
+    fd_vote_accounts_pair_serializable_t_map_insert( new_stakes->vote_accounts.vote_accounts_pool, &new_stakes->vote_accounts.vote_accounts_root, new_node );
+
+  }
+
+  /* Stale stake delegations should also be removed or updated in the cache. */
```

Comment:
(no need to change, just a note:) we will probably get rid of this stake delegations cache, as it's unnecessary, has caused a bunch of bugs and when the stake program is migrated to bpf we will not be able to update it inline. 

we will probably just keep an index of stake/vote accounts and re-compute the delegation values out of funk.